[interp] Add wide data item index static load and stores

lambdageek opened the pull request:
Add `ldsdfld.w` and `stsfld.w` opcodes to the interpreter.  These are needed for some machine-generated methods in Android projects that use AndroidX, MAUI, SkiaSharp or other toolkits.  The app's `Resource::UpdateIdValues` can grow to exceed 64k interpreter data items, which means that 16-bit data item indices will roll over.

A typical snippet from the method body:

```
    IL_8413:  ldc.i4     0x7f0f02fc
    IL_8418:  stsfld     int32 [Microsoft.Maui.Controls.Compatibility]Microsoft.Maui.Controls.Compatibility.Resource/Style::Widget_MaterialComponents_Toolbar_Surface
    IL_841d:  ldc.i4     0x7f0f02fd
    IL_8422:  stsfld     int32 [Microsoft.Maui.Controls.Compatibility]Microsoft.Maui.Controls.Compatibility.Resource/Style::Widget_MaterialComponents_Tooltip
    IL_8427:  ldc.i4     0x7f0f02fe
    IL_842c:  stsfld     int32 [Microsoft.Maui.Controls.Compatibility]Microsoft.Maui.Controls.Compatibility.Resource/Style::Widget_Support_CoordinatorLayout
    IL_8431:  ldsfld     int32[] TestApp.Resource/Styleable::ActionBar
    IL_8436:  stsfld     int32[] [Microsoft.Maui.Controls.Compatibility]Microsoft.Maui.Controls.Compatibility.Resource/Styleable::ActionBar
    IL_843b:  ldsfld     int32[] TestApp.Resource/Styleable::ActionBarLayout
    IL_8440:  stsfld     int32[] [Microsoft.Maui.Controls.Compatibility]Microsoft.Maui.Controls.Compatibility.Resource/Styleable::ActionBarLayout
    IL_8445:  ldc.i4.0
    IL_8446:  stsfld     int32 [Microsoft.Maui.Controls.Compatibility]Microsoft.Maui.Controls.Compatibility.Resource/Styleable::ActionBarLayout_android_layout_gravity
```

The "wide" opcodes use 32-bit data item indices.

Resolves https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1398069

lambdageek committed:
warn if there's a ton of data items

lambdageek committed:
[interp] Add stsfld.vt.w interpreter opcode

For methods that require more than 64k data items, add a variant stsfld
instruction that stores data item indices as 32-bit integers instead of 16-bit.

To solve this problem in general, we need wide variants of more opcodes, but
for specific (machine generated) methods, a wide stsfld instruction is sufficient.

Resolves https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1398069

lambdageek committed:
Add ldsfld.vt.w opcode

lambdageek committed:
[mono] Log debug message when running .cctors

lambdageek committed:
drop hacky logging

lambdageek commented:
For .NET 7.0 we should investigate how to fix this in general - I made an issue https://github.com/dotnet/runtime/issues/59221.

This version that just deals with `UpdateIdValues` methods should be backported to .NET 6.0

lambdageek committed:
fix whitespace

BrzVlad commented on src/mono/mono/mini/interp/interp.c:
```diff
@@ -5418,6 +5418,18 @@ MINT_IN_CASE(MINT_BRTRUE_I8_SP) ZEROP_SP(gint64, !=); MINT_IN_BREAK;
+			gpointer addr = frame->imethod->data_items [READ32 (ip + 4)];
+			guint16 size = ip [6];
+
+			memcpy (locals + ip [1], addr, size);
```
This is incorrect because extension of the type is not happening here. For example loading an int8 needs to write the extended int32 to the local var. This can be done with the slow but generic stackval_from_data.

BrzVlad commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -1098,6 +1098,30 @@ get_data_item_index (TransformData *td, void *ptr)
+get_data_item_wide_index (TransformData *td, void *ptr)
+{
+	gpointer p = g_hash_table_lookup (td->data_hash, ptr);
+	uint32_t index;
```
There is quite a bit of duplicated code here. Why not just have a `get_data_item_index_wide` and the old `get_data_item_index` which calls `get_data_item_index_wide` and then asserts the index is <= G_MAXUINT16.

BrzVlad commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -4189,17 +4224,25 @@ interp_emit_sfld_access (TransformData *td, MonoClassField *field, MonoClass *fi
 				td->last_ins->data [0] = get_data_item_index (td, field_class);
 		}
 	} else {
+		gboolean wide_data = is_data_item_wide_index (td);
```
Logically there shouldn't be this side effect for other opcodes. I think it would make more sense to just get the required indexes with `get_data_item_index_wide` and if both indexes are short use the standard opcodes otherwise use the wide version.

BrzVlad commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -4208,15 +4251,26 @@ interp_emit_sfld_access (TransformData *td, MonoClassField *field, MonoClass *fi
+			if (mt == MINT_TYPE_VT)
+				td->last_ins->data [2] = size;
+		} else {
+			uint32_t vtable_idx = get_data_item_wide_index (td, vtable);
```
I would prefer to use the glib defines instead of the stdint ones for the sake of consistency with the rest of the code in the interpreter.

BrzVlad commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -1107,10 +1131,18 @@ get_data_item_index (TransformData *td, void *ptr)
 	index = td->n_data_items;
 	td->data_items [index] = ptr;
 	++td->n_data_items;
+	g_assertf (index < (UINT32_MAX - 1), "Interpreter data item 32-bit wide index 0x%x for method '%s' overflows", index, td->method->name);
```
I think it is pointless to assert for 32bit limit. Nothing will ever work on interp for such methods, not just the data items.

lambdageek committed:
Address review feedback

Use stackval_from_data and stackval_to_data instead of memcpy.

Don't duplicate code get_data_item_wide_index.

Only use the wide opcodes if any of the data item indices of a given
instruction need it, not just when the method total number of indices
overflows.

lambdageek commented:
Ran some local tests always using the wide opcodes and it seems like it's working.   But I'd appreciate another look at the `stackval_from_data`/`stackval_to_data` in interp.c @BrzVlad

BrzVlad commented on src/mono/mono/mini/interp/mintops.def:
```diff
@@ -85,6 +85,7 @@ OPDEF(MINT_LDSFLD_R4, "ldsfld.r4", 4, 1, 0, MintOpUShortInt)
 OPDEF(MINT_LDSFLD_R8, "ldsfld.r8", 4, 1, 0, MintOpUShortInt)
 OPDEF(MINT_LDSFLD_O, "ldsfld.o", 4, 1, 0, MintOpUShortInt)
 OPDEF(MINT_LDSFLD_VT, "ldsfld.vt", 5, 1, 0, MintOpTwoShorts)
+OPDEF(MINT_LDSFLD_VT_W, "ldsfld.vt.w", 8, 1, 0, MintOpTwoInts)
```
We use the VT suffix for opcodes operating on MINT_TYPE_VT. We should drop it for these opcodes.

BrzVlad commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -1111,6 +1111,20 @@ get_data_item_index (TransformData *td, void *ptr)
+get_data_item_index (TransformData *td, void *ptr)
+{
+	guint32 index = get_data_item_wide_index (td, ptr);
+	g_assertf (index < (UINT16_MAX - 1), "Interpreter data item index 0x%x for method '%s' overflows", index, td->method->name);
```
I don't understand the reasoning for this comparison. Why not simply `index <= G_MAXUINT16` ? Since provided index must just fit in an guint16 slot.

lambdageek replied:
Eh. you're right. For some reason the fact that they're stored as `GUINT_TO_POINTER (index + 1)` in the hash table was throwing me off.  But that doesn't matter since it will be wider there.

BrzVlad commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -1111,6 +1111,20 @@ get_data_item_index (TransformData *td, void *ptr)
+static gboolean
+is_data_item_wide_index (guint32 data_item_index)
+{
+	return data_item_index >= (gint32)(UINT16_MAX - 1);
```
Same as above

BrzVlad reviewed (approved):
Looks good. Just those last nitpicks

lambdageek committed:
Fix wide index boundary conditions

lambdageek committed:
Rename opcodes to ldsfld.w and stsfld.w

lambdageek commented:
/backport to release/6.0-rc2
